### ISSUE TITLE
rotate the wtmp/btmp logs timeout after 122.096s.

### ISSUE THREAD

**Issue by xiaozhu36** [cloudfoundry/bosh-linux-stemcell-builder#97]
When I use the [Alibaba Cloud Bosh CPI](https://github.com/cloudfoundry-incubator/bosh-alicloud-cpi-release) and the latest [Alibaba Cloud stemcell](https://bosh.io/stemcells/bosh-alicloud-kvm-ubuntu-xenial-go_agent) to run the smoke test, there always happens a timeouts error:
```
• Failure [147.178 seconds]
Stemcell when logrotate wtmp/btmp logs [It] should rotate the wtmp/btmp logs 
/tmp/build/82c27cda/bosh-linux-stemcell-builder/src/github.com/cloudfoundry/stemcell-acceptance-tests/ipv4director/smoke/smoke_test.go:34

  Timed out after 122.096s.
  Logfile '/var/log/wtmp' was larger than expected. It should have been rotated.
```
Even if I shorten the interval from 15 seconds to 5 seconds, it also failed after several retry:
```
default/76a7fe46-6c8b-41c7-b1ad-aa001ee1e624: stderr | Connection to 172.16.0.4 closed.
[Cmd Runner] 2019/07/22 19:24:03 DEBUG - Stderr: 
[Cmd Runner] 2019/07/22 19:24:03 DEBUG - Successful: true (0)
[Cmd Runner] 2019/07/22 19:24:03 DEBUG - Running command '/usr/local/bin/bosh -n -d stemcell-acceptance-tests ssh --column=stdout --results default/0 sudo du /var/log/wtmp | cut -f1'
[Cmd Runner] 2019/07/22 19:24:14 DEBUG - Stdout: 4	
	
[Cmd Runner] 2019/07/22 19:24:14 DEBUG - Stderr: 
[Cmd Runner] 2019/07/22 19:24:14 DEBUG - Successful: true (0)
[Cmd Runner] 2019/07/22 19:24:19 DEBUG - Running command '/usr/local/bin/bosh -n -d stemcell-acceptance-tests ssh --column=stdout --results default/0 sudo du /var/log/wtmp | cut -f1'
[Cmd Runner] 2019/07/22 19:24:29 DEBUG - Stdout: 4	
	
[Cmd Runner] 2019/07/22 19:24:29 DEBUG - Stderr: 
[Cmd Runner] 2019/07/22 19:24:29 DEBUG - Successful: true (0)
[Cmd Runner] 2019/07/22 19:24:34 DEBUG - Running command '/usr/local/bin/bosh -n -d stemcell-acceptance-tests ssh --column=stdout --results default/0 sudo du /var/log/wtmp | cut -f1'
[Cmd Runner] 2019/07/22 19:24:47 DEBUG - Stdout: 4	
	
[Cmd Runner] 2019/07/22 19:24:47 DEBUG - Stderr: 
[Cmd Runner] 2019/07/22 19:24:47 DEBUG - Successful: true (0)
[Cmd Runner] 2019/07/22 19:24:52 DEBUG - Running command '/usr/local/bin/bosh -n -d stemcell-acceptance-tests ssh --column=stdout --results default/0 sudo du /var/log/wtmp | cut -f1'
[Cmd Runner] 2019/07/22 19:25:02 DEBUG - Stdout: 8	
	
[Cmd Runner] 2019/07/22 19:25:02 DEBUG - Stderr: 
[Cmd Runner] 2019/07/22 19:25:02 DEBUG - Successful: true (0)
[Cmd Runner] 2019/07/22 19:25:07 DEBUG - Running command '/usr/local/bin/bosh -n -d stemcell-acceptance-tests ssh --column=stdout --results default/0 sudo du /var/log/wtmp | cut -f1'
[Cmd Runner] 2019/07/22 19:25:18 DEBUG - Stdout: 8	
	
[Cmd Runner] 2019/07/22 19:25:18 DEBUG - Stderr: 
[Cmd Runner] 2019/07/22 19:25:18 DEBUG - Successful: true (0)
[Cmd Runner] 2019/07/22 19:25:23 DEBUG - Running command '/usr/local/bin/bosh -n -d stemcell-acceptance-tests ssh --column=stdout --results default/0 sudo du /var/log/wtmp | cut -f1'
[Cmd Runner] 2019/07/22 19:25:33 DEBUG - Stdout: 12	
	
[Cmd Runner] 2019/07/22 19:25:33 DEBUG - Stderr: 
[Cmd Runner] 2019/07/22 19:25:33 DEBUG - Successful: true (0)
[Cmd Runner] 2019/07/22 19:25:38 DEBUG - Running command '/usr/local/bin/bosh -n -d stemcell-acceptance-tests ssh --column=stdout --results default/0 sudo du /var/log/wtmp | cut -f1'
[Cmd Runner] 2019/07/22 19:25:49 DEBUG - Stdout: 12	
	
[Cmd Runner] 2019/07/22 19:25:49 DEBUG - Stderr: 
[Cmd Runner] 2019/07/22 19:25:49 DEBUG - Successful: true (0)
[Cmd Runner] 2019/07/22 19:25:54 DEBUG - Running command '/usr/local/bin/bosh -n -d stemcell-acceptance-tests ssh --column=stdout --results default/0 sudo du /var/log/wtmp | cut -f1'

------------------------------
[Cmd Runner] 2019/07/22 19:26:05 DEBUG - Stdout: 12	
	
[Cmd Runner] 2019/07/22 19:26:05 DEBUG - Stderr: 
[Cmd Runner] 2019/07/22 19:26:05 DEBUG - Successful: true (0)
• Failure [147.178 seconds]
Stemcell when logrotate wtmp/btmp logs [It] should rotate the wtmp/btmp logs 
/tmp/build/82c27cda/bosh-linux-stemcell-builder/src/github.com/cloudfoundry/stemcell-acceptance-tests/ipv4director/smoke/smoke_test.go:34

  Timed out after 122.096s.
  Logfile '/var/log/wtmp' was larger than expected. It should have been rotated.
```
Is there any ideas to fix it?

In addition, this CI pipeline is [bosh-cpi-certification of alicloud](https://github.com/aliyun/bosh-cpi-certification/tree/master/alicloud).

**Comment by aashah** [comment]
Hey @xiaozhu36,

This test pertains to our logrotate rules described here - https://github.com/cloudfoundry/bosh-linux-stemcell-builder/blob/00a6d23ef2387b2fb1beb61c942bb41856915e9b/stemcell_builder/stages/logrotate_config/assets/ubuntu-logrotate.conf

Are the contents of `/var/lib/logrotate.status` consistent with the test failures, in that log rotation is not happening?

Are there any error logs related to `logrotate` that you can find on the smoke test VM? Here are some helpful links for getting that info: 
- https://access.redhat.com/solutions/32831

**Comment by xiaozhu36** [comment]
HI  @aashah Thanks for your feedback. According to your suggestion, I logged on the smoke test VM and run some commands, the results as following:

* There is no logrotate.staus when running command `cat /var/lib/logrotate.status`
   The commands shows `cat: /var/lib/logrotate.status: No such file or directory`

* There is no any error when running command `/usr/sbin/logrotate -d /etc/logrotate.conf`
   The commands shows:
```
reading config file /etc/logrotate.conf
including /etc/logrotate.d
reading config file apt
reading config file chrony
reading config file default_su_directive
reading config file dpkg
reading config file dracut-core
reading config file rsyslog
reading config file vcap

Handling 11 logs

rotating pattern: /var/log/apt/term.log  monthly (12 rotations)
empty log files are not rotated, old logs are removed
switching euid to 0 and egid to 108
considering log /var/log/apt/term.log
  log /var/log/apt/term.log does not exist -- skipping
switching euid to 0 and egid to 0

rotating pattern: /var/log/apt/history.log  monthly (12 rotations)
empty log files are not rotated, old logs are removed
switching euid to 0 and egid to 108
considering log /var/log/apt/history.log
  log /var/log/apt/history.log does not exist -- skipping
switching euid to 0 and egid to 0

rotating pattern: /var/log/chrony/*.log  weekly (7 rotations)
empty log files are not rotated, old logs are removed
switching euid to 0 and egid to 108
considering log /var/log/chrony/measurements.log
  log does not need rotating
considering log /var/log/chrony/statistics.log
  log does not need rotating
considering log /var/log/chrony/tracking.log
  log does not need rotating
not running postrotate script, since no logs were rotated
switching euid to 0 and egid to 0

rotating pattern: /var/log/dpkg.log  monthly (12 rotations)
empty log files are not rotated, old logs are removed
considering log /var/log/dpkg.log
  log /var/log/dpkg.log does not exist -- skipping

rotating pattern: /var/log/alternatives.log  monthly (12 rotations)
empty log files are not rotated, old logs are removed
considering log /var/log/alternatives.log
  log /var/log/alternatives.log does not exist -- skipping

rotating pattern: /var/log/dracut.log  monthly (12 rotations)
empty log files are not rotated, old logs are removed
considering log /var/log/dracut.log
  log /var/log/dracut.log does not exist -- skipping

rotating pattern: /var/log/syslog
 5242880 bytes (7 rotations)
empty log files are not rotated, old logs are removed
switching euid to 104 and egid to 108
considering log /var/log/syslog
  log does not need rotating
switching euid to 0 and egid to 0

rotating pattern: /var/log/mail.info
/var/log/mail.warn
/var/log/mail.err
/var/log/mail.log
/var/log/daemon.log
/var/log/kern.log
/var/log/auth.log
/var/log/user.log
/var/log/lpr.log
/var/log/cron.log
/var/log/debug
/var/log/messages
/var/log/news/news.crit
/var/log/news/news.err
/var/log/news/news.notice
 5242880 bytes (4 rotations)
empty log files are not rotated, old logs are removed
switching euid to 104 and egid to 108
considering log /var/log/mail.info
  log /var/log/mail.info does not exist -- skipping
considering log /var/log/mail.warn
  log /var/log/mail.warn does not exist -- skipping
considering log /var/log/mail.err
  log /var/log/mail.err does not exist -- skipping
considering log /var/log/mail.log
  log /var/log/mail.log does not exist -- skipping
considering log /var/log/daemon.log
  log does not need rotating
considering log /var/log/kern.log
  log does not need rotating
considering log /var/log/auth.log
  log does not need rotating
considering log /var/log/user.log
  log /var/log/user.log does not exist -- skipping
considering log /var/log/lpr.log
  log /var/log/lpr.log does not exist -- skipping
considering log /var/log/cron.log
  log /var/log/cron.log does not exist -- skipping
considering log /var/log/debug
  log /var/log/debug does not exist -- skipping
considering log /var/log/messages
  log /var/log/messages does not exist -- skipping
considering log /var/log/news/news.crit
  log /var/log/news/news.crit does not exist -- skipping
considering log /var/log/news/news.err
  log /var/log/news/news.err does not exist -- skipping
considering log /var/log/news/news.notice
  log /var/log/news/news.notice does not exist -- skipping
not running postrotate script, since no logs were rotated
switching euid to 0 and egid to 0

rotating pattern: /var/vcap/data/sys/log/*.log /var/vcap/data/sys/log/.*.log /var/vcap/data/sys/log/*/*.log /var/vcap/data/sys/log/*/.*.log /var/vcap/data/sys/log/*/*/*.log /var/vcap/data/sys/log/*/*/.*.log  52428800 bytes (7 rotations)
empty log files are rotated, old logs are removed
considering log /var/vcap/data/sys/log/*.log
  log /var/vcap/data/sys/log/*.log does not exist -- skipping
considering log /var/vcap/data/sys/log/.*.log
  log /var/vcap/data/sys/log/.*.log does not exist -- skipping
considering log /var/vcap/data/sys/log/bosh-dns/bosh_dns_ctl.stderr.log
  log does not need rotating
considering log /var/vcap/data/sys/log/bosh-dns/bosh_dns_ctl.stdout.log
  log does not need rotating
considering log /var/vcap/data/sys/log/bosh-dns/bosh_dns_health_ctl.stderr.log
  log does not need rotating
considering log /var/vcap/data/sys/log/bosh-dns/bosh_dns_health_ctl.stdout.log
  log does not need rotating
considering log /var/vcap/data/sys/log/bosh-dns/bosh_dns_health.stderr.log
  log does not need rotating
considering log /var/vcap/data/sys/log/bosh-dns/bosh_dns_health.stdout.log
  log does not need rotating
considering log /var/vcap/data/sys/log/bosh-dns/bosh_dns_resolvconf_ctl.stderr.log
  log does not need rotating
considering log /var/vcap/data/sys/log/bosh-dns/bosh_dns_resolvconf_ctl.stdout.log
  log does not need rotating
considering log /var/vcap/data/sys/log/bosh-dns/bosh_dns_resolvconf.stderr.log
  log does not need rotating
considering log /var/vcap/data/sys/log/bosh-dns/bosh_dns_resolvconf.stdout.log
  log does not need rotating
considering log /var/vcap/data/sys/log/bosh-dns/bosh_dns.stderr.log
  log does not need rotating
considering log /var/vcap/data/sys/log/bosh-dns/bosh_dns.stdout.log
  log does not need rotating
considering log /var/vcap/data/sys/log/bosh-dns/post-start.stderr.log
  log does not need rotating
considering log /var/vcap/data/sys/log/bosh-dns/post-start.stdout.log
  log does not need rotating
considering log /var/vcap/data/sys/log/bosh-dns/pre-start.stderr.log
  log does not need rotating
considering log /var/vcap/data/sys/log/bosh-dns/pre-start.stdout.log
  log does not need rotating
considering log /var/vcap/data/sys/log/syslog_forwarder/pre-start.stderr.log
  log does not need rotating
considering log /var/vcap/data/sys/log/syslog_forwarder/pre-start.stdout.log
  log does not need rotating
considering log /var/vcap/data/sys/log/*/.*.log
  log /var/vcap/data/sys/log/*/.*.log does not exist -- skipping
considering log /var/vcap/data/sys/log/syslog_forwarder/blackbox/blackbox.stderr.log
  log does not need rotating
considering log /var/vcap/data/sys/log/syslog_forwarder/blackbox/blackbox.stdout.log
  log does not need rotating
considering log /var/vcap/data/sys/log/*/*/.*.log
  log /var/vcap/data/sys/log/*/*/.*.log does not exist -- skipping

rotating pattern: /var/log/wtmp  5242880 bytes (4 rotations)
empty log files are not rotated, old logs are removed
considering log /var/log/wtmp
  log does not need rotating

rotating pattern: /var/log/btmp  5242880 bytes (4 rotations)
empty log files are not rotated, old logs are removed
considering log /var/log/btmp
  log does not need rotating
```
In addition, this error does not always occur and it is a random error.

**Comment by aashah** [comment]
Sorry for responding so late, we'll do better!

Our suspicions are moving towards cron, and why logrotate is not happening on a regular interval. Can we get some information about crontab:

- is cron enabled: `$ systemctl status cron.service`
- confirming `logrotate` is happening: '$ grep CRON /var/log/syslog | grep logrotate`

**Comment by xiaozhu36** [comment]
HI @aashah Thanks for your reply. Please check the following commands results:
```
$ systemctl status cron.service
● cron.service - Regular background program processing daemon
   Loaded: loaded (/lib/systemd/system/cron.service; enabled; vendor preset: enabled)
   Active: active (running) since Tue 2019-07-23 07:08:52 UTC; 3 weeks 1 days ago
     Docs: man:cron(8)
 Main PID: 531 (cron)
   CGroup: /system.slice/cron.service
           └─531 /usr/sbin/cron -f

Aug 15 04:35:01 39825bd0-cd34-4bbc-ba55-2a0d0d194e37 CRON[14561]: pam_unix(cron:session): session opened for user root by (uid=0)
Aug 15 04:35:01 39825bd0-cd34-4bbc-ba55-2a0d0d194e37 CRON[14565]: (root) CMD (/usr/bin/logrotate-cron)
Aug 15 04:45:01 39825bd0-cd34-4bbc-ba55-2a0d0d194e37 CRON[14573]: pam_unix(cron:session): session opened for user root by (uid=0)
Aug 15 04:45:01 39825bd0-cd34-4bbc-ba55-2a0d0d194e37 CRON[14574]: (root) CMD (command -v debian-sa1 > /dev/null && debian-sa1 1 1)
Aug 15 04:45:01 39825bd0-cd34-4bbc-ba55-2a0d0d194e37 CRON[14573]: pam_unix(cron:session): session closed for user root
Aug 15 04:50:01 39825bd0-cd34-4bbc-ba55-2a0d0d194e37 CRON[14577]: pam_unix(cron:session): session opened for user root by (uid=0)
Aug 15 04:50:01 39825bd0-cd34-4bbc-ba55-2a0d0d194e37 CRON[14578]: (root) CMD (/usr/bin/logrotate-cron)
Aug 15 04:55:01 39825bd0-cd34-4bbc-ba55-2a0d0d194e37 CRON[14590]: pam_unix(cron:session): session opened for user root by (uid=0)
Aug 15 04:55:01 39825bd0-cd34-4bbc-ba55-2a0d0d194e37 CRON[14591]: (root) CMD (command -v debian-sa1 > /dev/null && debian-sa1 1 1)
Aug 15 04:55:01 39825bd0-cd34-4bbc-ba55-2a0d0d194e37 CRON[14590]: pam_unix(cron:session): session closed for user root
```
```
$ cat /etc/cron.d/logrotate
5,20,35,50 * * * * root /usr/bin/logrotate-cron
```
```
$ grep CRON /var/log/syslog | grep logrotate
2019-08-07T16:20:01.521561+00:00 localhost CRON[660]: (root) CMD (/usr/bin/logrotate-cron)
2019-08-07T16:35:01.551577+00:00 localhost CRON[675]: (root) CMD (/usr/bin/logrotate-cron)
2019-08-07T16:50:01.587412+00:00 localhost CRON[694]: (root) CMD (/usr/bin/logrotate-cron)
2019-08-07T17:05:01.617847+00:00 localhost CRON[708]: (root) CMD (/usr/bin/logrotate-cron)
2019-08-07T17:20:01.655826+00:00 localhost CRON[730]: (root) CMD (/usr/bin/logrotate-cron)
...
```
This issue will lower our customer's confidence and I need your help eagerly. Thanks in advance.

**Comment by xiaozhu36** [comment]
HI @aashah Is there any ideas for this issue?

**Comment by xiaozhu36** [comment]
HI @aashah I noticed this issue is a random error and not always timeout. Does it help you to find the reason?

**Comment by xiaozhu36** [comment]
HI @aashah Is there any ideas for this?

**Comment by einavle** [comment]
Hi @xiaozhu36,
In order to fix this, you need to increase the test timeout beyond 15 minutes. (20 minutes will satisfy)
You can see in the file: `setup-logrotate.sh` that the rotation intervals are 15 minutes, therefore, it is unlikely that the rotation will be executed within the test timeout period.

**Comment by xiaozhu36** [comment]
Hi @einavle Thanks for your reply. Actually, I have tested several times based on setting 15min and more, but the issue still exist. Now, I have submitted the PR #101 to fix this issue. After that, my pipeline always is in success.

HI @aashah Can you check the PR #101 ? This is very important for us. Thanks in advance.

**Comment by einavle** [comment]
@xiaozhu36,
I understand. I have also placed my PR for your review.
https://github.com/cloudfoundry/bosh-linux-stemcell-builder/pull/102

**Comment by xiaozhu36** [comment]
HI @einavle I made a test based on your PR, but sometimes it is still failed. So, in my PR, I added an extra returning condition that if the current log size is less than 5000, it should be success and return.

**Comment by xiaozhu36** [comment]
Fixed by #101
